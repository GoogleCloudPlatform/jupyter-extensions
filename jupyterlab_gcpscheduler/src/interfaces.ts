export interface Bucket {
  name: string;
  accessLevel?: 'uniform' | 'fine';
}

export interface Buckets {
  buckets: Bucket[];
}

export interface CloudStorageApiBucket {
  iamConfiguration?: {
    uniformBucketLevelAccess?: {
      /** If set, access is controlled only by bucket-level or above IAM policies. */
      enabled?: boolean;
    };
  };
  /** The ID of the bucket. For buckets, the id and name properties are the same. */
  id?: string;
}

export interface CloudStorageApiBuckets {
  /** The list of items. */
  items?: CloudStorageApiBucket[];
  /** The kind of item this is. For lists of buckets, this is always storage#buckets. */
  kind?: string;
  /** The continuation token, used to page through large result sets. Provide this value in a subsequent request to return the next page of results. */
  nextPageToken?: string;
}

/** Interfaces used by services to call and read Notebooks Executor API responses */

interface ApiClientObjectMap<T> {
  [key: string]: T;
}

interface Status {
  code?: number;
  message?: string;
}

export interface Operation {
  name?: string;
  metadata?: ApiClientObjectMap<any>;
  done?: boolean;
  error?: Status;
  response?: ApiClientObjectMap<any>;
}

export interface CreateScheduleResponse {
  error?: string;
}

export interface CreateExecutionResponse {
  error?: string;
}

export interface ListSchedulesResponse {
  schedules?: NotebooksApiSchedule[];
  nextPageToken?: string;
  unreachable?: string[];
}

export interface ListExecutionsResponse {
  executions?: NotebooksApiExecution[];
  nextPageToken?: string;
  unreachable?: string[];
}

type NotebooksApiExecutionState =
  | 'STATE_UNSPECIFIED'
  | 'QUEUED'
  | 'PREPARING'
  | 'RUNNING'
  | 'SUCCEEDED'
  | 'FAILED'
  | 'CANCELLING'
  | 'CANCELLED'
  | 'PAUSED';

type NotebooksApiScheduleState =
  | 'STATE_UNSPECIFIED'
  | 'ENABLED'
  | 'PAUSED'
  | 'DISABLED'
  | 'UPDATE_FAILED';

type NotebooksApiExecutionTemplateScaleTier =
  | 'SCALE_TIER_UNSPECIFIED'
  | 'BASIC'
  | 'STANDARD_1'
  | 'PREMIUM_1'
  | 'BASIC_GPU'
  | 'BASIC_TPU'
  | 'CUSTOM';

type NotebooksApiSchedulerAcceleratorConfigType =
  | 'SCHEDULER_ACCELERATOR_TYPE_UNSPECIFIED'
  | 'NVIDIA_TESLA_K80'
  | 'NVIDIA_TESLA_P100'
  | 'NVIDIA_TESLA_V100'
  | 'NVIDIA_TESLA_P4'
  | 'NVIDIA_TESLA_T4'
  | 'NVIDIA_TESLA_A100'
  | 'TPU_V2'
  | 'TPU_V3'
  | 'TPU_V2_POD'
  | 'TPU_V3_POD';

export interface NotebooksApiExecution {
  name?: string;
  displayName?: string;
  description?: string;
  createTime?: string;
  updateTime?: string;
  state?: NotebooksApiExecutionState;
  executionTemplate?: NotebooksApiExecutionTemplate;
  outputNotebookFile?: string;
}

export interface NotebooksApiSchedule {
  name?: string;
  displayName?: string;
  description?: string;
  state?: NotebooksApiScheduleState;
  cronSchedule?: string;
  timeZone?: string;
  createTime?: string;
  updateTime?: string;
  executionTemplate?: NotebooksApiExecutionTemplate;
}

export interface NotebooksApiExecutionTemplate {
  scaleTier?: NotebooksApiExecutionTemplateScaleTier;
  masterType?: string;
  acceleratorConfig?: NotebooksApiSchedulerAcceleratorConfig;
  labels?: ApiClientObjectMap<string>;
  inputNotebookFile?: string;
  containerImageUri?: string;
  location?: string;
  outputNotebookFolder?: string;
  paramsYamlFile?: string;
}

export interface NotebooksApiSchedulerAcceleratorConfig {
  type?: NotebooksApiSchedulerAcceleratorConfigType;
  coreCount?: string;
}

/** Interfaces used by extension to display and create Executions/Schedules */

/** Message type describing an AI Platform training Job */
export interface ExecuteNotebookRequest {
  imageUri: string;
  inputNotebookGcsPath: string;
  name: string;
  masterType: string;
  outputNotebookFolder: string;
  gcsBucket: string;
  scaleTier: string;
  region: string;
  acceleratorType: string;
  acceleratorCount: string;
}

export interface ExecutionTemplate {
  id: string;
  name: string;
  updateTime?: string;
  createTime?: string;
  gcsFile: string;
  state: string;
  link?: string;
  viewerLink?: string;
  downloadLink?: string;
  timeZone?: string;
}

/** UI interface used to represent a Scheduled Notebook Job */
export interface Execution extends ExecutionTemplate {
  type: string;
  bucketLink?: string;
}

export interface Schedule extends ExecutionTemplate {
  schedule: string;
  hasExecutions: boolean;
}

export interface Executions {
  executions: Execution[];
  pageToken: string;
}

export interface Schedules {
  schedules: Schedule[];
  pageToken: string;
}
