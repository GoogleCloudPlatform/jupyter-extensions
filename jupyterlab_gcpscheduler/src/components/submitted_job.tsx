/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { withStyles } from '@material-ui/core';
import { CheckCircle } from '@material-ui/icons';
import { COLORS, css, LearnMoreLink } from 'gcp_jupyterlab_shared';
import * as React from 'react';
import { classes, stylesheet } from 'typestyle';

import {
  findOptionByValue,
  REGIONS,
  SCALE_TIERS,
  GCS_LINK,
  SCHEDULER_LINK,
  MASTER_TYPES,
  CUSTOM,
  ACCELERATOR_TYPES,
} from '../data';
import { RunNotebookRequest } from '../service/gcp';
import { OnDialogClose } from './dialog';
import { ActionBar } from './action_bar';

interface Props {
  onFormReset: () => void;
  onDialogClose: OnDialogClose;
  request: RunNotebookRequest;
  projectId: string;
  schedule?: string;
}

interface GcsInformation {
  basename: string;
  link: string;
}

const localStyles = stylesheet({
  dl: {
    fontSize: '13px',
    $nest: {
      '& > div': {
        marginBottom: '8px',
      },
      '& dt': {
        display: 'inline',
        fontWeight: 500,
      },
      '& dd': {
        display: 'inline',
        marginLeft: '4px',
      },
    },
  },
  message: {
    alignItems: 'center',
    fontSize: '15px',
    fontWeight: 500,
  },
});

const GreenCheck = withStyles({
  root: {
    color: COLORS.green,
    fontSize: '28px',
    marginRight: '8px',
  },
})(CheckCircle);

export class SubmittedJob extends React.Component<Props, {}> {
  render() {
    const {
      onDialogClose,
      onFormReset,
      request,
      projectId,
      schedule,
    } = this.props;
    const projectParam = `project=${projectId}`;
    const region = findOptionByValue(REGIONS, request.region);
    const scaleTier = findOptionByValue(SCALE_TIERS, request.scaleTier);
    const masterType = findOptionByValue(MASTER_TYPES, request.masterType);
    const acceleratorType = findOptionByValue(
      ACCELERATOR_TYPES,
      request.acceleratorType
    );
    const gcsInformation = this.getGcsInformation(request.inputNotebookGcsPath);
    const gcsLink = `${GCS_LINK}/${gcsInformation.link}?${projectParam}`;
    const isRecurring = !!schedule;
    const jobLink = `${SCHEDULER_LINK}?${projectParam}`;
    return (
      <div className={css.column}>
        <div className={classes(css.row, localStyles.message)}>
          <GreenCheck />
          <span>Notebook run successfully scheduled!</span>
        </div>
        <dl className={localStyles.dl}>
          <div>
            <dt>Notebook:</dt>
            <dd>
              <LearnMoreLink text={gcsInformation.basename} href={gcsLink} />
            </dd>
          </div>
          <div>
            <dt>Run name:</dt>
            <dd>{request.jobId}</dd>
          </div>
          <div>
            <dt>Region:</dt>
            <dd>{region.text}</dd>
          </div>
          <div>
            <dt>Scale tier:</dt>
            <dd>
              {scaleTier.value} - {scaleTier.text}
            </dd>
          </div>
          {scaleTier.value === CUSTOM && (
            <React.Fragment>
              <div>
                <dt>Master type:</dt>
                <dd>{masterType.text}</dd>
              </div>
              <div>
                <dt>Accelerator type:</dt>
                <dd>{acceleratorType.text}</dd>
              </div>
              <div>
                <dt>Accelerator count:</dt>
                <dd>{request.acceleratorCount}</dd>
              </div>
            </React.Fragment>
          )}
          <div>
            <dt>Container:</dt>
            <dd>{request.imageUri}</dd>
          </div>
          <div>
            <dt>Type:</dt>
            <dd>{isRecurring ? 'Recurring run' : 'Single run'}</dd>
          </div>
          {isRecurring && (
            <div>
              <dt>Schedule:</dt>
              <dd>{schedule}</dd>
            </div>
          )}
          <div>
            <dt>Job:</dt>
            <dd>
              <LearnMoreLink text="View in Cloud Console" href={jobLink} />
            </dd>
          </div>
        </dl>
        <ActionBar closeLabel="Done" onDialogClose={onDialogClose}>
          <button className={css.button} type="button" onClick={onFormReset}>
            Submit another job
          </button>
        </ActionBar>
      </div>
    );
  }

  // Strips off the gs:// prefix and returns everything up to the file name
  private getGcsInformation(notebookPath: string): GcsInformation {
    const sliceEnd = notebookPath.lastIndexOf('/');
    return {
      basename: notebookPath.slice(sliceEnd + 1),
      link: notebookPath.slice('gs://'.length, sliceEnd),
    };
  }
}
