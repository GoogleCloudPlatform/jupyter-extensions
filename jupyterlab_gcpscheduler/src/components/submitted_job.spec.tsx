/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { shallow } from 'enzyme';
import { Button } from '@material-ui/core';
import { SubmittedJob } from './submitted_job';
import { TEST_PROJECT } from '../test_helpers';
import { ExecuteNotebookRequest } from '../interfaces';

class MockDate extends Date {
  constructor() {
    super('2020-11-11T15:00:00');
  }
}

describe('SubmittedJob', () => {
  const projectId = TEST_PROJECT;
  const gcsBucket = `gs://${TEST_PROJECT}`;
  const mockOnFormReset = jest.fn();
  const mockDialogClose = jest.fn();
  const request: ExecuteNotebookRequest = {
    name: 'test_execution_1234',
    imageUri: 'gcr.io/deeplearning-platform-release/tf-gpu.1-15:latest',
    inputNotebookGcsPath: `${gcsBucket}/test_execution_1234/test_execution_1234.ipynb`,
    masterType: 'n1-standard-4',
    outputNotebookFolder: `${gcsBucket}/test_execution/test_execution.ipynb`,
    scaleTier: 'CUSTOM',
    gcsBucket: 'gcsBucket',
    region: 'us-east1',
    acceleratorType: 'NVIDIA_TESLA_K80',
    acceleratorCount: '1',
  };
  const mockProps = {
    onDialogClose: mockDialogClose,
    onFormReset: mockOnFormReset,
    request,
    projectId,
    schedule: '',
  };

  const mockPropsWithCustomContainer = {
    onDialogClose: mockDialogClose,
    onFormReset: mockOnFormReset,
    request: { ...request, imageUri: 'gcr.io/test' },
    projectId,
    schedule: '',
  };

  beforeEach(() => {
    jest.resetAllMocks();
    global.Date = MockDate as any;
  });

  it('Renders for immediate job', async () => {
    const creator = shallow(<SubmittedJob {...mockProps} />);
    expect(creator).toMatchSnapshot();
  });

  it('Renders for immediate job with custom container', async () => {
    const creator = shallow(<SubmittedJob {...mockPropsWithCustomContainer} />);
    expect(creator).toMatchSnapshot();
  });

  it('Renders for scheduled job', async () => {
    mockProps.schedule = '0 15 * * *';
    const creator = shallow(<SubmittedJob {...mockProps} />);
    expect(creator).toMatchSnapshot();
  });

  it('Invokes onFormReset when Submit another job is clicked', async () => {
    const creator = shallow(<SubmittedJob {...mockProps} />);
    creator.find(Button).simulate('click');
    expect(mockOnFormReset).toHaveBeenCalled();
  });
});
