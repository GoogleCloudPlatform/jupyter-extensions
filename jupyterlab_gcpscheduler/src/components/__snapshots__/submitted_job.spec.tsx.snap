// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmittedJob Renders for immediate job 1`] = `
<div
  className="column_fghohlb"
>
  <div
    className="row_f17j33oz message_ftgw0r2"
  >
    <WithStyles(CheckCircleIcon) />
    <span>
      Notebook run successfully scheduled!
    </span>
  </div>
  <dl
    className="dl_f1xmw2no"
  >
    <div>
      <dt>
        Notebook:
      </dt>
      <dd>
        <LearnMoreLink
          href="https://console.cloud.google.com/storage/browser/test-project/test_job_1234?project=test-project"
          text="test_job_1234.ipynb"
        />
      </dd>
    </div>
    <div>
      <dt>
        Run name:
      </dt>
      <dd>
        test_job_1234
      </dd>
    </div>
    <div>
      <dt>
        Region:
      </dt>
      <dd>
        us-east1 (South Carolina)
      </dd>
    </div>
    <div>
      <dt>
        Scale tier:
      </dt>
      <dd>
        CUSTOM
         - 
        Custom machine type configuration
      </dd>
    </div>
    <div>
      <dt>
        Master type:
      </dt>
      <dd>
        4 CPUs, 15 GB RAM
      </dd>
    </div>
    <div>
      <dt>
        Accelerator type:
      </dt>
      <dd>
        NVIDIA Tesla K80
      </dd>
    </div>
    <div>
      <dt>
        Accelerator count:
      </dt>
      <dd>
        1
      </dd>
    </div>
    <div>
      <dt>
        Container:
      </dt>
      <dd>
        gcr.io/deeplearning-platform-release/tf-gpu.1-15:latest
      </dd>
    </div>
    <div>
      <dt>
        Type:
      </dt>
      <dd>
        Single run
      </dd>
    </div>
    <div>
      <dt>
        Job:
      </dt>
      <dd>
        <LearnMoreLink
          href="https://console.cloud.google.com/ai-platform/notebooks/list/scheduled-runs?project=test-project"
          text="View in Cloud Console"
        />
      </dd>
    </div>
  </dl>
  <ActionBar
    closeLabel="Done"
    onDialogClose={[MockFunction]}
  >
    <button
      className="button_f23yboz"
      onClick={[MockFunction]}
      type="button"
    >
      Submit another job
    </button>
  </ActionBar>
</div>
`;

exports[`SubmittedJob Renders for scheduled job 1`] = `
<div
  className="column_fghohlb"
>
  <div
    className="row_f17j33oz message_ftgw0r2"
  >
    <WithStyles(CheckCircleIcon) />
    <span>
      Notebook run successfully scheduled!
    </span>
  </div>
  <dl
    className="dl_f1xmw2no"
  >
    <div>
      <dt>
        Notebook:
      </dt>
      <dd>
        <LearnMoreLink
          href="https://console.cloud.google.com/storage/browser/test-project/test_job_1234?project=test-project"
          text="test_job_1234.ipynb"
        />
      </dd>
    </div>
    <div>
      <dt>
        Run name:
      </dt>
      <dd>
        test_job_1234
      </dd>
    </div>
    <div>
      <dt>
        Region:
      </dt>
      <dd>
        us-east1 (South Carolina)
      </dd>
    </div>
    <div>
      <dt>
        Scale tier:
      </dt>
      <dd>
        CUSTOM
         - 
        Custom machine type configuration
      </dd>
    </div>
    <div>
      <dt>
        Master type:
      </dt>
      <dd>
        4 CPUs, 15 GB RAM
      </dd>
    </div>
    <div>
      <dt>
        Accelerator type:
      </dt>
      <dd>
        NVIDIA Tesla K80
      </dd>
    </div>
    <div>
      <dt>
        Accelerator count:
      </dt>
      <dd>
        1
      </dd>
    </div>
    <div>
      <dt>
        Container:
      </dt>
      <dd>
        gcr.io/deeplearning-platform-release/tf-gpu.1-15:latest
      </dd>
    </div>
    <div>
      <dt>
        Type:
      </dt>
      <dd>
        Recurring run
      </dd>
    </div>
    <div>
      <dt>
        Schedule:
      </dt>
      <dd>
        0 15 * * *
      </dd>
    </div>
    <div>
      <dt>
        Job:
      </dt>
      <dd>
        <LearnMoreLink
          href="https://console.cloud.google.com/ai-platform/notebooks/list/scheduled-runs?project=test-project"
          text="View in Cloud Console"
        />
      </dd>
    </div>
  </dl>
  <ActionBar
    closeLabel="Done"
    onDialogClose={[MockFunction]}
  >
    <button
      className="button_f23yboz"
      onClick={[MockFunction]}
      type="button"
    >
      Submit another job
    </button>
  </ActionBar>
</div>
`;
